feat(config): Config Verification Prebuild Step

Raxnam opened the pull request:
A prebuild script that checks for a valid platform config file before starting the build. This throws the error and stops the build before the build actually starts instead of after the build is done

Commit:
feat(config): Config Verification Prebuild Step

A prebuild script that checks for a valid platform config file before starting the build.
This throws the error and stops the build before the build actually starts instead of after the build is done

pew-dahs commented on Assets/Plugins/Source/Editor/PreProcessConfigConfirmation.cs:
```diff
@@ -0,0 +1,155 @@
+    public int callbackOrder { get { return 0; } }
+    public void OnPreprocessBuild(BuildReport report)
+    {
+        Debug.Log("PreProcessConfigConfirmation_Windows.OnPreprocessBuild for target " + report.summary.platform + " at path " + System.IO.Directory.GetCurrentDirectory());
```
I'm assuming this should be PreProcessingConfigConfirmation by itself instead of _Windows attached to it? Unless there's going to be platform specific pre process confirmations.

Raxnam replied:
Ah yea, I had initially planned to have one for each platform, but the preprocess function provides the target so everything can be done in one function.

pew-dahs commented on Assets/Plugins/Source/Editor/PreProcessConfigConfirmation.cs:
```diff
@@ -0,0 +1,155 @@
+
+        switch (target)
+        {
+            case BuildTarget.StandaloneWindows:
```
You can simplify this copy/pasted switch to instead use a readonly dictionary (or you could create a local one) of the target and the config file, something like
```
Dictionary<BuildTarget, string> buildFileMap = new Dictionary<BuildTarget, string> {
  { BuildTarget.StandaloneWindows, "Assets/StreamingAssets/EOS/EpicOnlineServicesConfig.json" },
  { BuildTarget.StandaloneWindows64, "Assets/StreamingAssets/EOS/EpicOnlineServicesConfig.json" },
  ...
}
```
Then the whole function can be simplified to
```
if (System.IO.File.Exists(buildFileMap[target]))
{
    Debug.Log("PreProcessConfigConfirmation.OnPreprocessBuild for target " + target + " config file exists");
}
else
{
    Debug.LogError("PreProcessConfigConfirmation.OnPreprocessBuild for target " + target + " config file is missing");
    throw new BuildFailedException("Config file is missing, refer to above message");
}
```
and you can remove the extra functions at the end. In this way, when we need to add new platforms, we don't have to copy/paste a section, but add a new line to the dictionary (or uncomment a line).

pew-dahs replied:
Also, you should have a default check for the case where the target platform isn't covered with some error message for that too.

Raxnam replied:
Thanks, I'm not too familiar with c# data structures, I just know it doesn't have a map, but that works really well.

pew-dahs replied:
A dictionary in C# is the same as a map in C++, just a different name.

Commit:
refactor(config confirmation) cleaned up code and added some default cases

Cleaned up the switch statement to just use a dictionary, and added a target is not supported case, along with clarifying some of the log messages

Commit:
refactor(config confirmation) removed unnecessary functions

removed the exists and does not exists function calls and replaced them with the error logging

pew-dahs commented on Assets/Plugins/Source/Editor/PreProcessConfigConfirmation.cs:
```diff
@@ -0,0 +1,70 @@
+        fileDictionary.Add(BuildTarget.StandaloneLinux64,   "Assets/StreamingAssets/EOS/eos_linux_config.json");
+        fileDictionary.Add(BuildTarget.Android,             "Assets/StreamingAssets/EOS/eos_android_config.json");
+        /* Unupported cases(Playstation and XBox)
+        file
```
You should remove this line.

pew-dahs commented on Assets/Plugins/Source/Editor/PreProcessConfigConfirmation.cs:
```diff
@@ -0,0 +1,70 @@
+        BuildTarget target = report.summary.platform;
+
+        Dictionary<BuildTarget, string> fileDictionary = new Dictionary<BuildTarget, string>();
+        fileDictionary.Add(BuildTarget.StandaloneWindows,   "Assets/StreamingAssets/EOS/EpicOnlineServicesConfig.json");
```
nit: You can actually initialize a dictionary directly without adding key/value pairs after. No need to change it, but nice to know about C# initialization.
```
Dictionary<BuildTarget, string> fileDictionary = new Dictionary<BuildTarget, string> {
  { BuildTarget.StandaloneWindows, "Assets/StreamingAssets/EOS/EpicOnlineServicesConfig.json" },
  ...
};
```

pew-dahs reviewed (approved):
Looks good. Just need to remove an extra line that's there.

Commit:
docs(config confirmation) Cleaned up comments
